src/cargo/sources/git/oxide.rs

```diff
@@ -0,0 +1,315 @@
+///
+/// This can affect the way we open it and the cost associated with it.
+pub enum OpenMode {
+    /// We need `git_binary` configuration as well for being able to see credential helpers
+    /// that are configured with the `git` installation itself.
+    /// However, this is slow on windows (~150ms) and most people won't need it as they use the
+    /// standard index which won't ever need authentication, so we only enable this when needed.
+    ForFetch,
+}
+
+impl OpenMode {
+    /// Sometimes we don't need to pay for figuring out the system's git installation, and this tells
+    /// us if that is the case.
+    pub fn needs_git_binary_config(&self) -> bool {
+        match self {
+            OpenMode::ForFetch => true,
+        }
+    }
+}
+
+/// Produce a repository with everything pre-configured according to `config`. Most notably this includes
+/// transport configuration. Knowing its `purpose` helps to optimize the way we open the repository.
+/// Use `config_overrides` to configure the new repository.
+pub fn open_repo(
+    repo_path: &std::path::Path,
+    config_overrides: Vec<BString>,
+    purpose: OpenMode,
+) -> Result<gix::Repository, gix::open::Error> {
+    gix::open_opts(repo_path, {
+        let mut opts = gix::open::Options::default();
+        opts.permissions.config = gix::permissions::Config::all();
+        opts.permissions.config.git_binary = purpose.needs_git_binary_config();
+        opts.with(gix::sec::Trust::Full)
+            .config_overrides(config_overrides)
+    })
+}
+
+/// Convert `git` related cargo configuration into the respective `git` configuration which can be
+/// used when opening new repositories.
+pub fn cargo_config_to_gitoxide_overrides(config: &Config) -> CargoResult<Vec<BString>> {
+    use gix::config::tree::{gitoxide, Core, Http, Key};
+    let timeout = HttpTimeout::new(config)?;
+    let http = config.http_config()?;
+
+    let mut values = vec![
+        gitoxide::Http::CONNECT_TIMEOUT.validated_assignment_fmt(&timeout.dur.as_millis())?,
+        Http::LOW_SPEED_LIMIT.validated_assignment_fmt(&timeout.low_speed_limit)?,
+        Http::LOW_SPEED_TIME.validated_assignment_fmt(&timeout.dur.as_secs())?,
+        // Assure we are not depending on committer information when updating refs after cloning.
+        Core::LOG_ALL_REF_UPDATES.validated_assignment_fmt(&false)?,
+    ];
+    if let Some(proxy) = &http.proxy {
+        values.push(Http::PROXY.validated_assignment_fmt(proxy)?);
+    }
+    if let Some(check_revoke) = http.check_revoke {
+        values.push(Http::SCHANNEL_CHECK_REVOKE.validated_assignment_fmt(&check_revoke)?);
+    }
+    if let Some(cainfo) = &http.cainfo {
+        values.push(
+            Http::SSL_CA_INFO.validated_assignment_fmt(&cainfo.resolve_path(config).display())?,
+        );
+    }
+    if let Some(user_agent) = &http.user_agent {
+        values.push(Http::USER_AGENT.validated_assignment_fmt(user_agent)?);
```

Comment:
Do we miss the default user agent `cargo {version}` here?

https://github.com/rust-lang/cargo/blob/39c13e67a5962466cc7253d41bc1099bbcb224c3/src/cargo/ops/registry.rs#L619

Reply:
🤦, that's a great catch and would have been quite a terrible bug if not caught! Thanks so much! It's fixed in 2b80c8f6e104a7ebe4bbdc477288840cdcc8278e and I chose a style similar to what's linked in `registry.rs`.